opm/models/utils/simulator.hh

```diff
@@ -99,6 +104,8 @@ class Simulator
     using MPIComm = typename Dune::MPIHelper::MPICommunicator;
     using Communication = Dune::Communication<MPIComm>;
 
+    static constexpr Scalar eps = constexpr_max<Scalar>(std::numeric_limits<Scalar>::epsilon(), 1.e-9);
```

Comment:
Passing an explicit template argument when the template parameter could be deduced from the input arguments are almost always a sign that we're doing something wrong.  Here, the problem is that `1.0e-9` is of type `double` which may not be the same as `Scalar`.  The typical way to fix this is to write
```C++
std::max(std::numeric_limits<Scalar>::epsilon, static_cast<Scalar>(1.0e-9))
```
instead.

Reply:
Thanks for pointing it out, I was thinking this way and did not realize using template parameter explicitly is something not encouraged.

Reply:
> did not realize using template parameter explicitly is something not encouraged.

Just to be clear, explicit template arguments are fine when they're needed.  `std::make_shared<>()` and `std::make_unique<>()` are obvious examples of this, because they need an explicit template argument to specify the pointer type.  It's just that passing explicit template arguments when the template parameter can be deduced from the "regular" function arguments is typically not a good idea.